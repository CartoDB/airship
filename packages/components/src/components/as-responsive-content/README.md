# as-responsive-content

This component manages the layout of the application. It searches for our layout elements and manages for you the responsiveness behaviour, creating the correspondant panels. See `Layout` section for more information on our layout elements.

## How to use it

Add a `<as-responsive-content>` element as a wrapper for the application sections.

The component searches for these classes in order to know how many tabs it must render:
- `as-main`.
- `as-sidebar`.
- `as-map-panels`.
- `as-map-footer`.

As long as you follow the proper markup, as the `Layout` section shows, you won't have to worry about anything else.

This component replaces `.as-content` HTML element, so it needs to wrap all your children elements to work properly. Here is an example:

```code
lang: html
---
<body class="as-app">
  <!-- Here goes your toolbar -->

  <as-responsive-content>
    <aside class="as-sidebar as-sidebar--left">
      LEFT-SIDEBAR
    </aside>

    <main class="as-main">
      <div class="as-map-area">
        <div class="as-map">
          <div id="map"></div>
        </div>

        <div class="as-map-footer">
          Map Footer
        </div>
      </div>
    </main>
  </as-responsive-content>
</body>
```

```hint|directive
Take into account that any layout dependent JavaScript code, like Leaflet or MapboxGL map, that use any elements inside the component will need to be executed after `ready` event is fired. Please check [`ready` event documentation](/developers/airship/reference/imported_content/catalog/#/components/application-content?a=ready).
```

### Setting the tab names

In order to set the tabs text, you need to add a data attribute with the desired name:

`<aside class="as-sidebar as-sidebar--left" data-name="Stores">`

`<div class="as-map-panels" data-name="Legends">`

`<aside class="as-sidebar as-sidebar--right" data-name="Containers">`

You can change that data attribute programatically if you need to update the responsive tab name.

### Example

```html
noSource: true
responsive: true
---
<iframe src="/examples/components/as-responsive-content/simple.html" style="width: 100%; height: 100%;">
```
[See example](/developers/airship/examples/#example-simple)

### Changing the tab order

If you want to change the tabs order, you can add a data attribute to tell the component their positions.

Let's put `Legends` in the first position, `Stores` in the second and `Containers` in third place.

```code
lang: html
showSource: false
---
<div class="as-main" data-tab-order="0">

<div class="as-map-panels" data-name="Legends" data-tab-order="1">

<aside class="as-sidebar as-sidebar--left" data-name="Stores" data-tab-order="2">

<aside class="as-sidebar as-sidebar--right" data-name="Containers" data-tab-order="3">
```

We recommend setting the map always at the position `0`.

```html
noSource: true
responsive: true
---
<iframe src="/examples/components/as-responsive-content/custom-order.html" style="width: 100%; height: 100%;">
```

[See example](/developers/airship/examples/#example-custom-order)

## Events

### **ready**

This event fires when the component has finished loading. Useful to load Leaflet with the proper content width and height.

```code
lang: javascript
---
const responsiveContent = document.querySelector('as-responsive-content');
responsiveContent.addEventListener('ready', () => {});
```

```hint|directive
To avoid errors with layout dependent JavaScript code, we recommend you to execute all your JavaScript code inside the ready event.
```

So, if you want to load a Leaflet map with some CARTO.js layers, you better put that code inside load event handler, like this:
```code
lang: javascript
---
const responsiveContent = document.querySelector('as-responsive-content');
responsiveContent.addEventListener('ready', () => {
  const map = L.map('map', {
    zoomControl: false
  }).setView([30, 0], 3);
  map.scrollWheelZoom.disable();

  L.tileLayer('https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png', {
    maxZoom: 18
  }).addTo(map);

  const client = new carto.Client({
    apiKey: 'default_public',
    username: 'cartojs-test'
  });

  const source = new carto.source.Dataset('ne_10m_populated_places_simple');
  const style = new carto.style.CartoCSS(`
    #layer {
      marker-width: 7;
      marker-fill: #EE4D5A;
      marker-line-color: #FFFFFF;
    }
  `);
  const layer = new carto.layer.Layer(source, style);

  client.addLayer(layer);
  client.getLeafletLayer().addTo(map);
});
```

### **sectionChanged**

This event fires when the content section is changed via clicking a tab or via the programmatic method.

```code
lang: javascript
---
const responsiveContent = document.querySelector('as-responsive-content');
responsiveContent.addEventListener('sectionChanged', (section) => { console.log(section); });
```

## Methods

### **getSections()**

This method return a list of the sections discovered throughout the content with some interesting properties, such as:

```code
lang: typescript
---
interface ApplicationSection {
  active?: boolean;
  activeAction?: (section: ApplicationSection) => {};
  activeClass?: string;
  disableAction?: (section: ApplicationSection) => {};
  element: HTMLElement;
  name: string;
  type: string;
  tabOrder: number;
}
```

```code
lang: javascript
---
// Async/Await approach
const responsiveContent = document.querySelector('as-responsive-content');
await responsiveContent.getSections();

// Promises approach
const responsiveContent = document.querySelector('as-responsive-content');
responsiveContent.getSections()
               .then(sections => console.log(sections));
```
```hint|directive
Please note that you always need to wrap your `await` code in an `async` function. If you use it outside of an async function, it will raise a `SyntaxError`. Learn more about it [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await).
```

### **setVisible(sectionName)**

This method changes visible content section by introducing the section name like:

```code
lang: javascript
showSource: false
---
// Async/Await approach
const responsiveContent = document.querySelector('as-responsive-content');
await responsiveContent.setVisible('sectionName');

// Promises approach
const responsiveContent = document.querySelector('as-responsive-content');
responsiveContent.setVisible()
                 .then(() => {
                   // Whatever you want to do next
                 });
```
```hint|directive
Please note that you always need to wrap your `await` code in an `async` function. If you use it outside of an async function, it will raise a `SyntaxError`. Learn more about it [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await).
```
