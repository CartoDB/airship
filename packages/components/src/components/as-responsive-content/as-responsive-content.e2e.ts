import { E2EElement, E2EPage, newE2EPage } from '@stencil/core/testing';

describe('as-responsive-content', () => {
  let page: E2EPage;

  beforeEach(async () => {
    page = await newE2EPage({ html: `<as-responsive-content>${domExample}</as-responsive-content>` });
  });

  describe('Rendering', async () => {
    it('should render tabs based on content', async () => {
      const tabs: E2EElement = await page.find('.as-toolbar-tabs');
      const tabsHTML = tabs.innerHTML;

      expect(tabsHTML).toContain('Sidebar 0');
      expect(tabsHTML).toContain('Sidebar 1');
      expect(tabsHTML).toContain('Panel 0');
      expect(tabsHTML).toContain('Bottom Bar');
    });
  });

  describe('Behaviour', async () => {
    let tabs: E2EElement[];
    beforeEach(async () => {
      tabs = await page.findAll('.as-tabs__item');
    });
    it('should display the map-area by default', async () => {
      const mapAreaElement = await page.find('.as-map-area');
      expect(mapAreaElement).toHaveClass('as-map-area--visible');
    });

    it('should activate a sidebar when clicking on its tab', async () => {
      const leftSidebar = await page.find('.as-sidebar--left');
      tabs[1].click();

      await page.waitForChanges();
      expect(leftSidebar).toHaveClass('as-sidebar--visible');
    });

    it('should activate map panels when clicking on its tab', async () => {
      const panels = await page.find('.as-map-panels');
      tabs[3].click();

      await page.waitForChanges();
      expect(panels).toHaveClass('as-map-panels--visible');
    });

    it('should activate the footer when clicking on its tab', async () => {
      const footer = await page.find('.as-map-footer');
      tabs[4].click();

      await page.waitForChanges();
      expect(footer).toHaveClass('as-map-footer--visible');
    });
  });
});

const domExample = `
  <aside class="as-sidebar as-sidebar--left">Left Sidebar</aside>

  <main class="as-main">
    <div class="as-map-area">
      <div id="map"></div>
      <div class="as-map-panels">
        <div class="as-panel as-panel--top as-panel--right">
          <div class="as-panel__element">Floating Panel</div>
        </div>
      </div>
    </div>

    <div class="as-map-footer">Bottom Bar</div>
  </main>

  <aside class="as-sidebar as-sidebar--right">Right Sidebar</aside>
`;
