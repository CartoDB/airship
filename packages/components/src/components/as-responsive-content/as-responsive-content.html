<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <base href="/">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
    <title>Airship</title>

    <!-- CARTO.js -->
    <!-- Include Leaflet -->
    <script src="https://unpkg.com/leaflet@1.3.1/dist/leaflet.js"></script>
    <link href="https://unpkg.com/leaflet@1.3.1/dist/leaflet.css" rel="stylesheet">
    <link href="https://carto.com/developers/carto-js/examples/maps/public/style.css" rel="stylesheet">
    <!-- Include CARTO.js -->
    <script src="https://libs.cartocdn.com/carto.js/v4.1.2/carto.min.js"></script>

    <link rel="stylesheet" href="packages/styles/dist/airship.css">
    <script type="module" src="packages/components/dist/airship/airship.esm.js"></script>
    <script nomodule="" src="packages/components/dist/airship/airship.js"></script>
  </head>

  <body class="as-app-body as-app">
    <as-responsive-content>
      <aside class="as-sidebar as-sidebar--left">
        LEFT-SIDEBAR
      </aside>

      <main class="as-main">
        <div class="as-map-area">
          <div class="as-map">
            <div id="map">
              MAP
            </div>

            <div class="as-map-panels">
              <div class="as-panel as-panel--top as-panel--right">
                <div class="as-panel__element">
                  <p>Title</p>
                  <p>
                    Lorem ipsum dolor sit amet consectetur adipisicing elit. Inventore possimus nam quam necessitatibus omnis, est sequi rem
                    quibusdam molestiae, at saepe odit voluptatem quae debitis consequatur iste ducimus quaerat dolorum.
                  </p>
                </div>
                <div class="as-panel__element"> PANEL 1 </div>
              </div>

              <div class="as-panel as-panel--middle as-panel--right as-panel--vertical">
                <div class="as-panel__element">
                  <p>Title</p>
                  <p>
                    Lorem ipsum dolor sit amet consectetur adipisicing elit. Inventore possimus nam quam necessitatibus omnis, est sequi rem
                    quibusdam molestiae, at saepe odit voluptatem quae debitis consequatur iste ducimus quaerat dolorum.
                  </p>
                </div>
                <div class="as-panel__element"> PANEL 1 </div>
              </div>
            </div>
          </div>
        </div>

        <footer class="as-map-footer">
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Explicabo delectus ullam laudantium necessitatibus pariatur repellendus
          nemo ipsam minus. Delectus quidem sequi recusandae culpa deleniti molestias amet consequatur aspernatur eos exercitationem.
        </footer>
      </main>

      <aside class="as-sidebar as-sidebar--right">
        RIGHT-SIDEBAR
      </aside>
    </as-responsive-content>

    <script type="text/javascript">
      const responsiveContent = document.querySelector('as-responsive-content');
      responsiveContent.addEventListener('load', () => {
        const map = L.map('map', {
          zoomControl: false
        }).setView([30, 0], 3);
        map.scrollWheelZoom.disable();

        L.tileLayer('https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png', {
          maxZoom: 18
        }).addTo(map);

        const client = new carto.Client({
          apiKey: 'default_public',
          username: 'cartojs-test'
        });

        const source = new carto.source.Dataset('ne_10m_populated_places_simple');
        const style = new carto.style.CartoCSS(`
                #layer {
                  marker-width: 7;
                  marker-fill: #EE4D5A;
                  marker-line-color: #FFFFFF;
                }
              `);
        const layer = new carto.layer.Layer(source, style);

        client.addLayer(layer);
        client.getLeafletLayer().addTo(map);
      });
    </script>
  </body>
</html>
