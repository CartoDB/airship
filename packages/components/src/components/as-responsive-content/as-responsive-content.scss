@import '../../../../styles/src/core/_dev';

as-responsive-content {
  display: flex;
  flex: 1;
  flex-direction: column;
}
