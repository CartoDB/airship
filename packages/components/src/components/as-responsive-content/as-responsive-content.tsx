import { Component, Element, Event, EventEmitter, Method, State } from '@stencil/core';
import { redrawChildren } from '../../utils/redraw-children';

/**
 * Category Widget
 *
 * @export
 * @class CategoryWidget
 */
@Component({
  shadow: false,
  styleUrl: './as-responsive-content.scss',
  tag: 'as-responsive-content'
})
export class ResponsiveContent {
  @Element() private element: HTMLElement;

  @State() private sections: ApplicationSection[] = [];

  @Event() private ready: EventEmitter<void>;
  @Event() private sectionChange: EventEmitter<object>;

  private activeSection: ApplicationSection;

  public render() {
    return [
      this.renderTabs(),
      this.renderContent()
    ];
  }

  public renderContent() {
    return (
      <section class='as-content'>
        <slot />
      </section>
    );
  }

  public componentWillLoad() {
    this.sections = this.getContentSections();
  }

  public componentDidLoad() {
    this.ready.emit();
  }

  @Method()
  public async getSections(): Promise<object[]> {
    return this.sections;
  }

  @Method()
  public async setVisible(sectionName: string) {
    const sectionFound = this.sections.find((section) => section.name === sectionName);

    if (sectionFound) {
      this.setActive(sectionFound);
    }
  }

  private renderTabs() {
    const tabs = this.sections.map((section, index) => {
      if (!section.element) {
        return;
      }

      const cssClasses = {
        'as-tabs__item': true,
        'as-tabs__item--active': section.active
      };

      return (
        <button onClick={() => this.setActive(section)} role='tab' class={cssClasses}>
          {section.name || index}
        </button>
      );
    });

    return (
      <div role='tablist' class='as-toolbar-tabs as-tabs as-tabs--xl'>{tabs}</div>
    );
  }

  private setActive(section: ApplicationSection) {
    if (section.active) {
      return;
    }

    this.disableActiveSection();

    if (section.activeAction) {
      section.activeAction(section);
    } else {
      section.element.classList.add(section.activeClass || ACTIVE_CLASSES[section.type]);
      redrawChildren(section.element);
      section.active = true;
    }

    this.activeSection = section;
    this.sections = [...this.sections];
    this.sectionChange.emit(section);
  }

  private disableActiveSection() {
    if (!this.activeSection) {
      return;
    }

    if (this.activeSection.disableAction) {
      this.activeSection.disableAction(this.activeSection);
      return;
    }

    this.activeSection.element.classList.remove(
      this.activeSection.activeClass || ACTIVE_CLASSES[this.activeSection.type]
    );
    this.activeSection.active = false;
  }

  private getContentSections() {
    const sections = [
      this.getMap(),
      ...this.getSidebars(),
      ...this.getPanels(),
      this.getMapFooter()
    ];

    if (sections.length) {
      sections.sort((a, b) => a.tabOrder - b.tabOrder);
      this.setActive(sections[0]);
    }

    return sections;
  }

  private getMap() {
    const mapWrapper = this.element.querySelector('.as-map-area');

    function activeAction(section: ApplicationSection) {
      section.active = true;
    }

    function disableAction(section: ApplicationSection) {
      section.active = false;
    }

    return {
      activeAction,
      disableAction,
      element: mapWrapper,
      name: mapWrapper.getAttribute('data-name') || 'Map',
      tabOrder: mapWrapper.getAttribute('data-tab-order') || 0,
      type: 'map'
    } as ApplicationSection;
  }

  private getSidebars() {
    const sidebars = Array.from(this.element.querySelectorAll('.as-sidebar'));

    const sidebarSections = sidebars.map((sidebar, index) => {
      return {
        activeClass: `as-sidebar--visible`,
        element: sidebar,
        name: sidebar.getAttribute('data-name') || `Sidebar ${index}`,
        tabOrder: sidebar.getAttribute('data-tab-order') || 0,
        type: 'sidebar'
      };
    }) as ApplicationSection[];

    return sidebarSections;
  }

  private getPanels() {
    const panels = Array.from(this.element.querySelectorAll('.as-map-panels'));

    const panelsSections = panels.map((panel, index) => ({
        element: panel,
        name: panel.getAttribute('data-name') || `Panel ${index}`,
        tabOrder: panel.getAttribute('data-tab-order') || 0,
        type: 'panels'
      })
    ) as ApplicationSection[];

    return panelsSections;
  }

  private getMapFooter() {
    const mapFooter = this.element.querySelector('.as-map-footer');

    return {
      element: mapFooter,
      name: mapFooter && mapFooter.getAttribute('data-name') || 'Bottom Bar',
      tabOrder: mapFooter && mapFooter.getAttribute('data-tab-order') || 0,
      type: 'mapFooter'
    } as ApplicationSection;
  }
}

const ACTIVE_CLASSES = {
  mapFooter: 'as-map-footer--visible',
  panels: 'as-map-panels--visible'
};

interface ApplicationSection {
  active?: boolean;
  activeAction?: (section: ApplicationSection) => {};
  activeClass?: string;
  disableAction?: (section: ApplicationSection) => {};
  element: HTMLElement;
  name: string;
  type: string;
  tabOrder: number;
}
