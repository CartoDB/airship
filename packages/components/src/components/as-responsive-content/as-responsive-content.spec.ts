import { TestWindow } from '@stencil/core/dist/testing';
import { ResponsiveContent } from './as-responsive-content';

describe('as-responsive-content', () => {
  it('should build', () => {
    expect(new ResponsiveContent()).toBeTruthy();
  });

  describe('Rendering', async () => {
    let element: HTMLAsResponsiveContentElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [ResponsiveContent],
        html: `<as-responsive-content>${domExample}</as-responsive-content>`
      });

      await testWindow.flush();
    });

    it('should render properly', () => {
      expect(element).toMatchSnapshot();
    });

    it('should render tabs based on content', () => {
      const tabs = element.querySelector('.as-toolbar-tabs');
      const tabsHTML = tabs.innerHTML;

      expect(tabsHTML).toContain('Sidebar 0');
      expect(tabsHTML).toContain('Sidebar 1');
      expect(tabsHTML).toContain('Panel 0');
      expect(tabsHTML).toContain('Bottom Bar');
    });

    it('should render child content inside .as-content', () => {
      const asAppContentNode = element.querySelector('.as-content');
      expect(asAppContentNode.innerHTML).toContain(domExample);
    });
  });

  describe('Behaviour', () => {
    let element: HTMLAsResponsiveContentElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [ResponsiveContent],
        html: `<as-responsive-content>${domExample}</as-responsive-content>`
      });

      await testWindow.flush();
    });

    it('should change active content when clicking tabs', async () => {
      const tabs = element.querySelectorAll('.as-tabs__item');

      let mapTab = tabs[0];
      expect(mapTab).toMatchClasses(['as-tabs__item', 'as-tabs__item--active']);

      let leftSidebarTab = element.querySelectorAll('.as-tabs__item')[1] as HTMLElement;
      leftSidebarTab.click();
      await testWindow.flush();

      const tabsAfterClick = element.querySelectorAll('.as-tabs__item');
      const leftSidebarNode = element.querySelector('.as-sidebar.as-sidebar--left');

      mapTab = tabsAfterClick[0];
      leftSidebarTab = tabsAfterClick[1] as HTMLElement;

      expect(mapTab).toMatchClasses(['as-tabs__item']);
      expect(leftSidebarTab).toMatchClasses(['as-tabs__item', 'as-tabs__item--active']);
      expect(leftSidebarNode).toMatchClasses(['as-sidebar', 'as-sidebar--left', 'as-sidebar--visible']);
    });
  });

  describe('Events', async () => {
    let element: HTMLAsResponsiveContentElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [ResponsiveContent],
        html: `<as-responsive-content>${domExample}</as-responsive-content>`
      });
    });

    it('should emit a "sectionChange" event when content changes in mobile view', async () => {
      const sectionChangeSpy = jest.fn();
      element.addEventListener('sectionChange', sectionChangeSpy);

      const leftSidebarTab = element.querySelectorAll('.as-tabs__item')[1] as HTMLElement;
      leftSidebarTab.click();

      await testWindow.flush();

      expect(sectionChangeSpy).toHaveBeenCalled();
    });
  });
});

const domExample = `
  <aside class="as-sidebar as-sidebar--left">Left Sidebar</aside>

  <main class="as-main">
    <div class="as-map-area">
      <div id="map"></div>
      <div class="as-map-panels">
        <div class="as-panel as-panel--top as-panel--right">
          <div class="as-panel__element">Floating Panel</div>
        </div>
      </div>
    </div>

    <div class="as-map-footer">Bottom Bar</div>
  </main>

  <aside class="as-sidebar as-sidebar--right">Right Sidebar</aside>
`;
