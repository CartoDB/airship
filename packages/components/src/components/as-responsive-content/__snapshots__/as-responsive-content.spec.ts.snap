// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-responsive-content Rendering should render properly 1`] = `
<as-responsive-content
  class="hydrated"
>
  
  
  
  
  
  
  
  
  <div
    class="as-toolbar-tabs as-tabs as-tabs--xl"
    role="tablist"
  >
    <button
      class="as-tabs__item as-tabs__item--active"
      role="tab"
    >
      Map
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Sidebar 0
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Sidebar 1
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Panel 0
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Bottom Bar
    </button>
  </div>
  <section
    class="as-content"
  >
    
    
  
    <aside
      class="as-sidebar as-sidebar--left"
    >
      Left Sidebar
    </aside>
    

  
    <main
      class="as-main"
    >
      
    
      <div
        class="as-map-area"
      >
        
      
        <div
          id="map"
        />
        
      
        <div
          class="as-map-panels"
        >
          
        
          <div
            class="as-panel as-panel--top as-panel--right"
          >
            
          
            <div
              class="as-panel__element"
            >
              Floating Panel
            </div>
            
        
          </div>
          
      
        </div>
        
    
      </div>
      

    
      <div
        class="as-map-footer"
      >
        Bottom Bar
      </div>
      
  
    </main>
    

  
    <aside
      class="as-sidebar as-sidebar--right"
    >
      Right Sidebar
    </aside>
    

  </section>
</as-responsive-content>
`;
